### ISSUE TITLE
Does scikit-fem support a tensor field?

### ISSUE THREAD

**Issue by adtzlr** [kinnala/scikit-fem#619]
Hi,

I do have a feature request. Would it be possible that scikit-fem supports a simple creation of a tensor field in the future? I mean something like

`skfem.element.element_vector.ElementTensor` (and `skfem.element.element_vector.ElementTensorSymmetric` as an optional extension...)

based on the already implemented `skfem.element.element_vector.ElementVector` (`skfem.element.ElementVectorH1`) class? I think I can use 3 seperate vector fields als column vectors of a tensor but the above suggestion would greatly simplify the creation of such fields.

I would like to create such a field with 3x3=9 components in 3d space with

`skfem.element.ElementTensor(skfem.element.ElementH1())`

I hope i did not overlook something regarding that feature in the docs.

Thanks in advance!

**Comment by gdmcbain** [comment]
It doesn't yet but it does seem like an obvious extension. Would you like to have a go?

**Comment by adtzlr** [comment]
I think I can't handle that in a way I would like to. In my opinion my solution is not ready for a pull request. Here is what I did: I copied `skfem/element/element_vector.py` to `skfem/element/element_tensor.py`. Then I basically replaced all `self.dim` by `self.dim**2`. 

The full code of this file looks like this:

```python
import numpy as np
from .element import Element
from .discrete_field import DiscreteField


class ElementTensor(Element):

    def __init__(self, elem):
        self.elem = elem

        # multiplicate number of dofs
        self.nodal_dofs = self.elem.nodal_dofs * self.dim**2
        self.facet_dofs = self.elem.facet_dofs * self.dim**2
        self.interior_dofs = self.elem.interior_dofs * self.dim**2
        self.edge_dofs = self.elem.edge_dofs * self.dim**2

        # new names for the vector components
        self.dofnames = [i + "^" + str(j + 1) + str(k + 1)
                         for i in elem.dofnames
                         for j in range(self.dim)
                         for k in range(self.dim)]
        self.maxdeg = elem.maxdeg
        self.refdom = elem.refdom

        if hasattr(elem, 'doflocs'):
            self.doflocs = np.array([
                elem.doflocs[int(np.floor(float(i) / float(self.dim**2)))]
                for i in range(self.dim**2 * elem.doflocs.shape[0])
            ])

    @property
    def dim(self):
        return self.elem.dim

    def gbasis(self, mapping, X, i, tind=None):
        """Set correct components to zero based on ``i``."""
        ind = int(np.floor(float(i) / float(self.dim**2)))
        n = i - self.dim**2 * ind
        fields = []
        for field in self.elem.gbasis(mapping, X, ind, tind)[0]:
            if field is None:
                fields.append(None)
            else:
                tmp = np.zeros((self.dim**2,) + field.shape)
                tmp[n] = field
                fields.append(tmp)
        return (DiscreteField(*fields),)
```

With some additions to `__init__.py` now this kind of code is possible: 

`skfem.element.ElementTensorH1(skfem.element.ElementHex1())`. 

My problem is that the tensor field is flattened, that means in 3d-space it has 9 components. I have to use a `fields(w)` function to extract and reshape the tensor fields. Additionaly, `np.reshape` does not work well with more than 1 undefined trailing axes. That means I have to explicitely get the shape information. I did it that way:

```python
def fields(w):
    _, a, b = w["sometensor"].value.shape
    
    u = w["displacements"]
    t = w["sometensor"].value.reshape(3,3,a,b)
    
    return u, t, a, b
```

and then

```python
def F2(w):
    u, t, a, b = fields(w)
    
    # this is just an example
    F = grad(u) + identity(u)

    return F - t
```

Inside the linear and bilinear forms I again have to reshape `u` and `v` for the tensor fields (`multiddot` is just a convenient helper which evaluates double-dot products of several arguments).

```python
@skfem.BilinearForm
def b22(u, v, w):
    _,a,b = u.value.shape
    t = u.value.reshape(3,3,a,b)
    δt = v.value.reshape(3,3,a,b)
    return multiddot(t, A22(w), δt)
```

If I could manage to avoid that the user has to take care of `a` and `b` in `u.value.reshape(3,3,a,b)` then my addition would be ready for a pull request. But I don't see an option for that. Any ideas? 

Thanks!

**Comment by adtzlr** [comment]
If I implement the reshaping in my personal helpers `dot` and `ddot` then this could be avoided by checking the dimensions of `u`, `Aij(w)` and `v` inside the helpers. Or something like a helper with an optional argument could also be useful: `dot(A, b, reshape=True)`. Anyway I think this should not be pushed to the `master` branch in that way.

**Comment by kinnala** [comment]
Is this not suitable?
```python
In [1]: from skfem import *                                                     

In [2]: m = MeshHex().refined()                                                 

In [3]: e = ElementVector(ElementVector(ElementHex1()))                         

In [4]: e                                                                       
Out[4]: <skfem.element.element_vector.ElementVector at 0x7f0e795acdc0>

In [5]: basis = InteriorBasis(m, e)                                             

In [6]: from skfem.helpers import ddot                                          

In [7]: A = asm(BilinearForm(lambda u, v, w: ddot(u, v)), basis)                

In [8]: A                                                                       
Out[8]: 
<243x243 sparse matrix of type '<class 'numpy.float64'>'
	with 3087 stored elements in Compressed Sparse Row format>
```

**Comment by adtzlr** [comment]
@kinnala Thanks for your suggestion! However, I get an error in the last line of the following code snippet:

```python
import skfem

m = skfem.mesh.MeshHex()
e = skfem.element.ElementVectorH1(
        skfem.element.ElementVectorH1(skfem.element.ElementHex1())
        )

basis = skfem.InteriorBasis(m, e)

A = basis.zeros()
basis.interpolate(A)
```

Error Message:
```python
Traceback (most recent call last):

  File "C:\Users\Andreas\Documents\GitHub\scikit-fem\test3.py", line 18, in <module>
    basis.interpolate(A)

  File "C:\Users\Andreas\anaconda3\lib\site-packages\skfem\assembly\basis\basis.py", line 244, in interpolate
    fs.append(linear_combination(n, ref[n]))

  File "C:\Users\Andreas\anaconda3\lib\site-packages\skfem\assembly\basis\basis.py", line 227, in linear_combination
    self.basis[i][c][-1][n][j, k, l]

TypeError: 'NoneType' object is not subscriptable```

**Comment by kinnala** [comment]
This was probably fixed in #574

**Comment by adtzlr** [comment]
I just tried it with the `master` branch and the error is gone. Thanks!
